cpp/src/groupby/hash/compute_groupby.cu

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "compute_single_pass_aggs.hpp"
+#include "helpers.cuh"
+#include "sparse_to_dense_results.hpp"
+
+#include <cudf/detail/aggregation/aggregation.cuh>
+#include <cudf/detail/aggregation/result_cache.hpp>
+#include <cudf/detail/cuco_helpers.hpp>
+#include <cudf/detail/gather.hpp>
+#include <cudf/groupby.hpp>
+#include <cudf/null_mask.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+#include <cuco/static_set.cuh>
+
+#include <iterator>
+#include <memory>
+
+namespace cudf::groupby::detail::hash {
+/**
+ * @brief Computes and returns a device vector containing all populated keys in
+ * `map`.
+ */
+template <typename SetType>
+rmm::device_uvector<size_type> extract_populated_keys(SetType const& key_set,
```

Comment:
Is this used outside of this TU? If not then it can be wrapped into anonymous namespace.

Reply:
> Is this used outside of this TU?

Yes, not in this PR though. In #16619 it will take a template parameter thus goes into a separate TU as well.

Reply:
I'm willing to be corrected on this, but I don't *think* function templates have linkage associated with them. I also think their instantiations are deemed inline.

If the template is defined in the `.cpp`/`.cu` file, I don't think they need to be shielded within `using namespace {}`.

Godbolt verification: https://godbolt.org/z/xnW7q5e67

`cppreference` is reticent in its discussion of templates and linkage. I'm still looking.

Reply:
I think @ttnghia's focus was to make this function private to the current TU.

Reply:
If it will be called from a separate TU, I would expect to see the declaration for `extract_populated_keys` in a header file so that it can be included by that other TU. Maybe I missed it.
Basically the header file would contain this line:
```
template <typename SetType>
rmm::device_uvector<size_type> extract_populated_keys(SetType const& key_set,                                                      
                                                      size_type num_keys,
                                                      rmm::cuda_stream_view stream);
```
This will be enough for the other TU to compile. 
The link step will then resolve the 2 defined here by this TU as appropriate.
This way, no matter how many TUs call this function, only the 2 here will ever be compiled and the linker will match them up at link time. This is different than with a header template where every TU recompiles the 2 functions and the linker then (sometimes) remove the duplicates.

Reply:
> I think @ttnghia's focus was to make this function private to the current TU.

Yep, hence the discussion of linkage.  `namespace {}` would be to force internal linkage (to the current TU).  My point is that with a function template defined in a `.cpp`, you wouldn't have to.

But, please ignore.  It's nitpicking.

Reply:
I've moved the declaration of `extract_populated_keys` to the header and it's no longer in the anonymous namespace.

Thank you all for the input.